adapters/finch/src/main/scala/caliban/FinchAdapter.scala

```diff
@@ -78,22 +78,25 @@ object FinchAdapter extends Endpoint.Module[Task] {
     skipValidation: Boolean = false,
     enableIntrospection: Boolean = true
   )(implicit runtime: Runtime[R]): Endpoint[Task, Json :+: Json :+: CNil] =
-    post(stringBody :: header("content-type")) { (body: String, contentType: String) =>
-      if (contentType == "application/graphql") {
-        val request = GraphQLRequest(Some(body))
-        executeRequest(request, interpreter, skipValidation = skipValidation, enableIntrospection = enableIntrospection)
-      } else {
-        val result = for {
-          query <- ZIO.fromEither(parse(body).flatMap(_.as[GraphQLRequest]))
-          result <- createRequest(
-                     query,
-                     interpreter,
-                     skipValidation = skipValidation,
-                     enableIntrospection = enableIntrospection
-                   )
-        } yield result
-        runtime.unsafeRunToFuture(result).future
-      }
+    post(queryParams :: stringBodyOption :: header("content-type")) {
+      (queryRequest: GraphQLRequest, body: Option[String], contentType: String) =>
+        val request: GraphQLRequest = (queryRequest, body, contentType) match {
+          case (_, Some(bodyValue), "application/json") =>
+            parse(bodyValue).flatMap(_.as[GraphQLRequest]).getOrElse(GraphQLRequest())
+          case (_, Some(_), "application/graphql") =>
+            GraphQLRequest(body)
+          case (queryRequest, _, _) if queryRequest.query.isDefined =>
+            queryRequest
+          // treat unmatched content-type as same as None of body.
+          case _ =>
+            GraphQLRequest()
```

Comment:
If we return a `Task[GraphQLRequest]`, this could fail with an exception e.g. "Query was not found".

Reply:
Maybe fixed this.